Refactoring list detail views templates

fkang-pu opened the pull request:
<!-- markdownlint-disable-next-line first-line-heading -->
## Summary Change Description
- data tables in templates for list and detail views:
  - tried to use same (or similar) data table for different webpages if possible
  - data tables are in **/DataRepo/templates/DataRepo/includes** directory
  - use "out_df" for output dataframe in data table  for "includes"
  - set client-side pagination, vertical scroll, default data size per page
  - use htmlSorter for column value with hyperlink

- Changes made to protocol list and detail views
- Changes made to compound detail page, fixed the bug.

 ## Affected Issue Numbers

- Resolves #534 
- Resolves #535
- Resolves #536
- Resolves #537 
- Resolves #541 
- Resolves #566 

## Code Review Notes

- Look like there are a lot of code changes in this PR, but they are mainly due to re-organizing data tables, except for protocol and compound views.
-  Based on 1st round of reviews, implemented two separate views and templates for "animal_treatment" and "msrun_protocol" as suggested in #535 . 
- Sorting "infusate" column was very slow and didn't work sometimes, removed it. It might be due to "if conditions". But can sort by "tracers" column.

- .Added test code to `tests/views/test_protocol_views.py`. Deleted old test code for protocols from `test_views.py`.
- Some column display could be better. Will collect feedback and address later if it requires any change to Pandas dataframe.
```
e.g.
study list page: 
infusates (by displaying tracer group or tracers?)
add msrun protocols? (will need to change dataframes)
```
## Checklist

- [x] All issue requirements satisfied (or no linked issues)
- [x] [Linting passes](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting).
- [x] [Migrations created & committed *(or no model changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)
- Tests
  - [ ] [Tests implemented *(or no code changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
  - [ ] [All *multi_working* tag tests pass locally](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
  - [ ] All example load tests pass (remotely) *(or their failures predated and are unrelated to this branch)*
  - [ ] `@tag("multi_working")` added to newly passing test functions/classes *(or no newly passing tests)*
  - [ ] `@tag("multi_broken")`, `@tag("multi_unknown")`, or `@tag("multi_mixed")` removed from newly passing test functions/classes *(or no newly passing tests)*

hepcat72 commented on DataRepo/templates/DataRepo/animal_detail.html:
```diff
@@ -61,101 +61,9 @@ <h5><a href="{% url 'search_basic' 'Animal' 'id' 'iexact' animal.id 'fcirc' %}">
 <br>
 <div>
     <h5>Sample Data for the Animal:</h5>
```
NB. Your headers above tables appear to vary as to whether they have a colon at the end or not.  If it was me, they're already h5 tags, so the colons seem superfluous.  I also think that subsequent tables' headers should be a lesser header tag (like h6 or whatever).

hepcat72 commented on DataRepo/templates/DataRepo/compound_detail.html:
```diff
@@ -42,92 +42,18 @@ <h5><a href="{% url 'search_basic' 'MeasuredCompound' 'id' 'iexact' compound.id
 <br>
```
NB. I think that peakgroups are defined already by their measured compound, so the headers above like `View Peak Data of this measured compound` seem a bit redundant.  I think it would be sufficient to say `View Peak Data`.

hepcat72 commented on DataRepo/templates/DataRepo/compound_detail.html:
```diff
@@ -42,92 +42,18 @@ <h5><a href="{% url 'search_basic' 'MeasuredCompound' 'id' 'iexact' compound.id
-    </table>
+        <h5>Measurement Data for the Animals using <b>{{ compound.name }}</b> as a tracer:</h5>
+        <br>
+        <h5><a href="{% url 'search_basic' 'TracerCompound' 'name' 'iexact' compound.name 'peakgroups' %}">View Peak Group Data</a></h5>
+        <h5><a href="{% url 'search_basic' 'TracerCompound' 'name' 'iexact' compound.name 'peakdata' %}">View Peak Data</a></h5>
+        <h5><a href="{% url 'search_basic' 'Compound' 'name' 'iexact' compound.name 'fcirc' %}">View Fcirc Data</a></h5>
```
NB. Still could use buttons, but also, the measured compounds includes the tracers (so change that long link to "all"), and I think instead of this section and the one above it for "measured compounds" as separate, it could be streamlined together into:
```suggestion
    {% if measured %}
        <h5><a href="{% url 'search_basic' 'MeasuredCompound' 'id' 'iexact' compound.id 'peakgroups' %}">View All Peak Groups</a></h5>
        <h5><a href="{% url 'search_basic' 'MeasuredCompound' 'id' 'iexact' compound.id 'peakdata' %}">View All Peak Data</a></h5>
        <h5><a href="{% url 'search_basic' 'TracerCompound' 'name' 'iexact' compound.name 'peakgroups' %}">View Tracer Peak Groups</a></h5>
        <h5><a href="{% url 'search_basic' 'TracerCompound' 'name' 'iexact' compound.name 'peakdata' %}">View Tracer Peak Data</a></h5>
        <h5><a href="{% url 'search_basic' 'Compound' 'name' 'iexact' compound.name 'fcirc' %}">View Fcirc Data</a></h5>
    {% else %}
        No measurements of this compound have yet been loaded.
    {% endif %}
```

fkang-pu replied:
Tried a few layouts,  it seems to me it's better by having links as "bulleted list"  on left side.  Made changes to detail pages for Study, Animal, and Compound.

hepcat72 commented on DataRepo/templates/DataRepo/compound_detail.html:
```diff
@@ -42,92 +42,18 @@ <h5><a href="{% url 'search_basic' 'MeasuredCompound' 'id' 'iexact' compound.id
+        <h5><a href="{% url 'search_basic' 'TracerCompound' 'name' 'iexact' compound.name 'peakdata' %}">View Peak Data</a></h5>
+        <h5><a href="{% url 'search_basic' 'Compound' 'name' 'iexact' compound.name 'fcirc' %}">View Fcirc Data</a></h5>
+        <br>
+        <h5>List of Animals using <b>{{ compound.name }}</b> as a tracer compound:</h5>
```
NB.
```suggestion
        <h5>Animals with <b>{{ compound.name }}</b> Tracers</h5>
```

hepcat72 commented on DataRepo/templates/DataRepo/compound_list.html:
```diff
@@ -21,7 +21,12 @@ <h4>List of Compounds</h4>
         data-show-export="true"
-        data-export-types="['csv', 'txt', 'excel']">
+        data-export-types="['csv', 'txt', 'excel']"
+        data-height="1200"
```
Q. Can this be made dynamic somehow? Tables with less than 10 rows have a lof of empty space.

fkang-pu replied:
> Q. Can this be made dynamic somehow? Tables with less than 10 rows have a lof of empty space.

one solution is to remove data-height, add the following css style to each of affected template files instead. I made changes to a few template files. [ref](https://github.com/wenzhixin/bootstrap-table/issues/1360)
    ```
<style>
        .fixed-table-body {
	        max-height: 1200px;
        }
    </style>
```

fkang-pu replied:
The above workaround solved "empty space" issue, but lost "sticky header" feature with setting both "data-height" and "data-virtual-scroll". I need to do a bit more research.

fkang-pu replied:
Added JavaScript code to make table height dynamic for a few web pages sharing same table layout (Study list vs. Protocol detail; Animal list vs. Compound detail).  Mainly to keep table responsive, have fixed table headers with vertical scrolling. Fixed extra space issue.

hepcat72 commented on DataRepo/templates/DataRepo/includes/animal_list_stats_table.html:
```diff
@@ -0,0 +1,93 @@
+                <th data-filter-control="input" data-sortable="true" data-sorter="htmlSorter" data-field="Studies">Studies</th>
+                <th data-filter-control="select" data-sortable="true" data-field="Genotype">Genotype</th>
+                <th data-filter-control="input" data-field="Infusate">Infusate</th>
+                <th data-filter-control="input" data-sortable="true" data-sorter="htmlSorter" data-field="Compounds">Parent Compound(s) for Tracer(s)</th>
```
NB. Really wide column.
```suggestion
                <th data-filter-control="input" data-sortable="true" data-sorter="htmlSorter" data-field="Compounds">Tracer Compound(s)</th>
```

hepcat72 commented on DataRepo/templates/DataRepo/includes/animal_list_stats_table.html:
```diff
@@ -0,0 +1,93 @@
+                <th data-filter-control="input" data-field="Infusate">Infusate</th>
+                <th data-filter-control="input" data-sortable="true" data-sorter="htmlSorter" data-field="Compounds">Parent Compound(s) for Tracer(s)</th>
+                <th data-filter-control="input" data-sortable="true" data-sorter="htmlSorter" data-field="Tracers">Tracer(s)</th>
+                <th data-filter-control="input" data-sortable="true" data-field="Concentrations">Infusion Concentration(s) (mM)</th>
```
NB.
```suggestion
                <th data-filter-control="input" data-sortable="true" data-field="Concentrations">Tracer Concentration(s) (mM)</th>
```

hepcat72 commented on DataRepo/templates/DataRepo/includes/animal_list_stats_table.html:
```diff
@@ -0,0 +1,93 @@
+                <th data-filter-control="select" data-sortable="true" data-visible="false" data-field="Sex">Sex</th>
+                <th data-filter-control="input" data-sortable="true" data-visible="false" data-field="Diet">Diet</th>
+                <th data-filter-control="input" data-sortable="true" data-field="Feeding-Status">Feeding Status</th>
+                <th data-filter-control="select" data-sortable="true" data-field="Sample-Owners">Sample Owners</th>
```
NB. If sample owner == animal owner, then...
```suggestion
                <th data-filter-control="select" data-sortable="true" data-field="Sample-Owners">Owner(s)</th>
```

hepcat72 reviewed (commented):
I'm only half way through the review, but so far it's looking great. Only minor suggestions and questions. I'll finish this review on Monday.

hepcat72 reviewed (approved):
All of my issues are superficial.  They only relate to *appearance* and efficiency/re-usability, not content.  This is a good improvement.  Feel free to address my issues as you see fit.

lparsons reviewed (commented):
Overall, this is seems like improvements, but there are some areas that I think we could still improve upon. One thing that would help when reviewing and discussing would be to have smaller PRs with only one focused change per PR. With so many changes in one PR, there are a lot of unrelated changes to review and discuss.

## Study Deail page: 
While I like the idea of the static table height, having it drop off below the bottom of the page is  not ideal. Also, the summary at the top, and especially the "View X Data" links take a lot of room away from the table. I would suggest reworking the general layout of these pages to be a bit more compact at the top.

## Protocols list view:
I think it would be more intuatitve for researchers to see a menu item for "Treatments" and one for "Mass Spec Protocols". Putting both on the same page with static height tables looks a bit clumsy to me, but is OK for now. It would be nice, however, if the tables actually fit inside the viewports and didn't create unnecessary scroll bars.

## Protocol detail view:
It would be nice if the title of the page was "Animal Treatment - NAME" and not "Protocol Record - Name". Same with MSRun Protocols, I think it would be more intuatitve if the page title was "Mass Spec Protocol - NAME".

## Compound detail view:
Suggestion:

View Peak Groups of this measured compound => Citrate Peak Groups
View Peak Data of this measured compound => Citrate Peak Data

---

#### Citrate used as a Tracer

Peak Groups
Peak Data
FCirc data

List of Animals using citrate as a tracer compound:

fkang-pu commented:
Tried to use JavaScript to set table height dynamically, but didn't make it work.  Have to use two more templates for now, and pushes some changes before taking a day off today.

fkang-pu commented on DataRepo/templates/DataRepo/includes/animal_sample_msrun_table.html:
```diff
@@ -0,0 +1,105 @@
+                <th data-filter-control="select" data-sortable="true" data-field="Sample-Owner">Sample Owner</th>
+                <th data-filter-control="input" data-sortable="true" data-field="Sample-Date">Sample Date</th>
+                <th data-filter-control="input" data-sortable="true" data-field="Collect-Time-Minutes">Sample Collect Time(m)</th>
+                <th data-filter-control="select" data-sortable="true" data-field="MSRun-Owner">MSRun Owner</th>
```
Each row in this table is for a sample/per msrun, so MSRun Owner is correct.

fkang-pu commented:
Made changes based on reviews. Fixed the issues listed by @lparsons, and fixed most of them raised by @hepcat72 . Two issues require more research and can be revisited in a few weeks.

1. table with more than 12 columns:
   tried the following approaches, but none of them worked better than current layout (responsive table, client-side pagination, fixed header with vertical scrolling).
   e.g.
   - add `<br>` to column header would mess up the "header" in downloaded file
   - use Bootstrap grid and column classes to control the width for a wide column in responsive table
   - set column width using **Bootstrap-table** option, but seemed not working for table with many columns 

2. Protocol list page
   Keep current layout with two tables on the same page for now. Fixed extra empty space issue. Will evaluate again when more data are loaded.

hepcat72 commented on DataRepo/templates/DataRepo/protocol_detail.html:
```diff
@@ -34,3 +39,26 @@ <h4>Studies using this protocol for MSRuns:</h4>
+
+{% block js_extras %}
+    {{ block.super }}
+    <script>
+        let rowCount = {{ df|length }}
+        let vScroll = undefined
+        let tabHeight = undefined
+        if ( rowCount < 5 ) {
+            tabHeight = undefined
+            vScroll = undefined
+        } else if (rowCount >= 5 && rowCount < 10) {
+            tabHeight = 800
+            vScroll = true
+        } else {
+            tabHeight = 1000
+            vScroll = true
+        }
+
+        $('#study_list_stats').bootstrapTable({
+                height: tabHeight
+        })
+    </script>
```
You're not ultimately using vScroll here, though you are in the version of this code in compound_detail.html.  If that's not just an oversight, then you can remove the occurrences of vScroll in this code.  Alternatively, can the code be re-used so that it only exists in one place?

fkang-pu replied:
@hepcat72 Nice catch! Fixed now. Thanks! I was testing with or without vScroll value, and forgot to add back for final version. I'll keep the code in each template for now, since I am not sure what is the best solution for setting the height for total rows <10, especially for rows between 5 to 9. The row height changes in the table. I only have less than 10 studies and protocols in my dev database, and would like to see the effects with much large datasets before re-using the code.

lparsons reviewed (changes requested):
This has a lot improvement and we should merge it, presuming a few issues are addressed.

## Study Detail page
**Blocking**:
There is some issue with the template. I see `{ block.super }}` at the top of the page and all of the text in the table is bold, but only the headers should be bold, the body should be normal weight.

## Protocols list page
**Non-Blocking**:
I would prefer to see two links on the left menu, one for Animal Treatments and one for Mass Spec Protocols. They can both go to the same page if we don't want to create two lists (though I would prefer separate pages I think).

## Protocol detail view
**Blocking**:
The Heading in the HTML now reads "Animal Treatment - no treatment", but the page Title (shown in the browswer tab) still reads "Protocol Record - no treatment".

lparsons commented:
@fkang-pu I'm getting the following error on the home page. 

```
Traceback (most recent call last):
  File "/home/lparsons/miniconda3/envs/tracebase/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/home/lparsons/miniconda3/envs/tracebase/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/lparsons/Documents/projects/tracebase/DataRepo/views/nav.py", line 91, in home
    "card_foot_url": reverse("protocol_list"),
  File "/home/lparsons/miniconda3/envs/tracebase/lib/python3.8/site-packages/django/urls/base.py", line 86, in reverse
    return resolver._reverse_with_prefix(view, prefix, *args, **kwargs)
  File "/home/lparsons/miniconda3/envs/tracebase/lib/python3.8/site-packages/django/urls/resolvers.py", line 698, in _reverse_with_prefix
    raise NoReverseMatch(msg)

Exception Type: NoReverseMatch at /
Exception Value: Reverse for 'protocol_list' not found. 'protocol_list' is not a valid view function or pattern name.
```

fkang-pu commented:
@lparsons I totally forgot the protocol link from the home page. looks like I went down the rabbit hole. I guess I have to use two cards for protocols (animal treatments and msrun protocols) on home page, otherwise the users would be very confused. Any suggestion?

fkang-pu commented:
@lparsons another approach is to remove protocol card from the home page, since they are not super important for the stats, a quick link from the sidebar may be sufficient. What do you think?

lparsons commented:
I think it would be OK to have both, only one, or none as cards. Here's a quick sketch idea to include both:
![image](https://user-images.githubusercontent.com/645128/197773009-e776d8c7-2e72-4686-b6cf-6ec87151f583.png)

fkang-pu commented:
@lparsons The home page is fixed now. I tried a few layout of the cards on home page, and also tried to replace "advanced search" card with a button. Seems to me it looks better to keep "advanced search" as a card in the 3rd row, especially when resizing the page. So I opt to use this layout.

 p.s. I took the opportunity to move the tests for home views into `DataRepo/tests/views/test_home_views.py` and modified the code.

lparsons commented on DataRepo/templates/DataRepo/study_detail.html:
```diff
@@ -41,110 +41,16 @@ <h4>Study Record - {{ study.name }}</h4>
-        </tbody>
-    </table>
-    {% endif %}
+    <h6>Animal and Sample Data for the Study</6>
```
Blocking: This is what is causing the bold font throughout the table.

```suggestion
    <h6>Animal and Sample Data for the Study</h6>
```

fkang-pu replied:
@lparsons Nice catch! fixed now.

lparsons reviewed (changes requested):
Looking very nice Fan. There is one small, but important, issue on the study detail page (bad closing tag for `</h6>`.

Also, this isn't blocking, but the layout for the "Advanced Search" box is a bit "off". Right now we have the two rows of summary cards within another row. The Advanced Search card is in a top level row. This throws off the width of the grids inside. 

```html
<div class="row">
    <div class="row">  <div>Studies</div> ... </div>
    <div class="row">...</div>
</div>

<div class="row"> <div>Advanced Search</div> </div>
```

fkang-pu commented:
@lparsons Hopefully I fixed errors by now. I made myself a bit confused by trying different layouts. Wish I handled the code/template changes in several PRs, lessons learned! Now just try to get this PR done ASAP. Thanks for a few rounds of reviews.

lparsons reviewed (approved):
LGTM!
